### DISCUSSION TITLE
ch32-hal-template

### DISCUSSION THREAD

**Discussion by romainreignier** [ch32-rs/ch32-hal#26] [Ideas]
Hi @andelf I have seen the new repo [ch32-hal-template](https://github.com/ch32-rs/ch32-hal-template)].
I assume that it is to provide a `cargo-generate` template like [esp-template](https://github.com/esp-rs/esp-template) right?
Have you started working on it?
I might try to help you on this but I wanted to know if I should start from zero or if you have already some code and/or ideas on which options you want to provide?
- Should we ask to use embassy or not (but the Delay implementation without embassy is missing for some chips family)?
- Should we only ask for the chip family, the exact variant or only some well-known boards?

**Comment by romainreignier** [comment]
I went ahead and created a first version: https://github.com/romainreignier/ch32-hal-template
For now, embassy is used, we can see later if we want to make it optional.
The main.rs is the embassy_blink example. The LED GPIO needs to be adjusted.

EDIT: later, a ci pipeline should be setup to check the generated templates

**Comment by andelf** [comment]
Great to hear you want to work on this! 
I've invited you to the repo and the ch32-rs team. Feel free to do the initial push of [ch32-hal-template](https://github.com/ch32-rs/ch32-hal-template). 🤓

> Should we ask to use Embassy or not (but the Delay implementation without Embassy is missing for some chip families)?

Embassy should be optional, but it's fine to leave it as the default.

> Should we only ask for the chip family...

CH32V boards are almost all simple ones. In my opinion, specifying the chip family is enough, with default IO settings for official EVT boards (many official EVTs have LEDs as free jumper pins).

**Reply by romainreignier** [reply]
Thanks, I have pushed my first version, I will try to make embassy optional and maybe add some CI.
